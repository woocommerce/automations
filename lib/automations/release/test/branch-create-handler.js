const {
	gimmeOctokit,
	gimmeContext,
	gimmeApp,
	gimmeConfig,
} = require( '@testHelpers' );

describe( 'branch-create-handler for release pull request generation', () => {
	let octokit, context, app, config;
	beforeEach( () => {
		octokit = gimmeOctokit();
		context = gimmeContext( 'create' );
		config = gimmeConfig( 'release' );
		app = gimmeApp( 'release' );
	} );
	it( 'creates a pull request', async () => {
		await app( context, octokit, config );
		expect( octokit.search.issuesAndPullRequests ).toHaveBeenCalled();
		expect( octokit.repos.getContents ).toHaveBeenCalled();
		expect(
			octokit.issues.listMilestonesForRepo.endpoint.merge
		).toHaveBeenCalled();
		expect( octokit.paginate.iterator ).toHaveBeenCalled();
		expect( octokit.issues.listForRepo.endpoint.merge ).toHaveBeenCalled();
		expect( octokit.paginate.iterator ).toHaveBeenCalled();
		expect( octokit.pulls.create ).toHaveBeenCalled();
		expect( octokit.issues.createComment ).toHaveBeenCalled();
		expect( octokit.pulls.create.mock.calls[ 0 ] ).toMatchSnapshot();
		expect(
			octokit.issues.createComment.mock.calls[ 0 ]
		).toMatchSnapshot();
	} );
	it( 'bails if pull branch is not a release branch', async () => {
		context.payload = { ...context.payload, ref: 'not-a-release-branch' };
		await app( context, octokit, config );
		expect( octokit.search.issuesAndPullRequests ).not.toHaveBeenCalled();
	} );
	describe( 'bails if pull request already exists for branch (various version strings)', () => {
		it.each`
			version
			${'3'}
			${'3.0'}
		`(
			'bails when the release branch is "release/$version" and there is already a pull with "Release: 3.0.0" as the title',
			async () => {
				octokit.search.issuesAndPullRequests.mockReturnValueOnce(
					Promise.resolve( {
						data: {
							items: [
								{
									title: 'Release: 3.0.0',
								},
							],
						},
					} )
				);
				context.payload = { ...context.payload, ref: 'release/3.0' };
				await app( context, octokit, config );
				expect(
					octokit.search.issuesAndPullRequests
				).toHaveBeenCalled();
				expect( octokit.repos.getBranch ).not.toHaveBeenCalled();
			}
		);
	} );
	it( 'Bails if there is no milestone for the given release branch', async () => {
		context.payload = { ...context.payload, ref: 'release/4.2.1' };
		await app( context, octokit, config );
		expect( octokit.pulls.create.mock.calls[ 0 ] ).toMatchSnapshot();
		expect(
			octokit.issues.createComment.mock.calls[ 0 ]
		).toMatchSnapshot();
	} );
} );
