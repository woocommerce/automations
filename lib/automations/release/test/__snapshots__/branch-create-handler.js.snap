// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`branch-create-handler for release pull request generation Bails if there is no milestone for the given release branch 1`] = `
Array [
  Object {
    "base": Promise {},
    "body": "This is the release pull request for WooCommerce Blocks plugin 4.2.1.

## Communication

>  This section is for any notes related to communicating the release Please include any extra details with each item as needed.


This release introduces:


> In this section document an overview/summary of what this release includes. You can refer to the changelog for more information


### Changelog

Changelog Error: Cannot find milestone by title: 4.2.1
You&#x27;ll need to edit this section manually

### Prepared Updates

The following documentation, blog posts, and changelog updates are prepared for the release:


> In this section you are highlighting all the public facing documentation that is related to the release. Feel free to remove anything that doesn't apply for this release.


**Release announcement:** *Link to release announcement post on developer.woocommerce.com (published after release)*

**Developer Notes** - The following issues require developer notes in the release post:

Devnotes Error: Cannot find milestone by title: 4.2.1
PRs tagged for dev notes cannot be found, you&#x27;ll need to edit this section manually.


**Relevant developer documentation:**
_Link(s) to any developer documentation related to the release_

**Happiness Engineer:**
_Link to any special instructions or helpful notes for HE related to this release_

* [ ] The release includes a changelog entry in the readme.txt?


## Quality

> This section is for any notes related to quality around the release Please include any extra details with each item as needed. This can include notes about why something isn't checked or expanding info on your response to an item.

**For testing:**

* [ ] Changes in this release are covered by Automated Tests.

      > This section is for confirming that the release changeset is covered by automated tests. If not, please leave some details on why not and any relevant information indicating confidence without those tests.

     * [ ] Unit tests
     * [ ] E2E tests
     * [ ] for each supported WordPress and WooCommerce core versions.

* This release has been tested on the following platforms:
     * [ ] mobile
     * [ ] desktop

* [ ] This release affects public facing REST APIs.
    * [ ] It conforms to REST API versioning policy.

* [ ] This release impacts **other extensions** or **backward compatibility**.
    * [ ] The release changes the signature of public methods or functions
        * [ ] This is documented (see: *Enter a link to the documentation here*)
    * [ ] The release affects filters or action hooks.
        * [ ] This is documented (see: *Enter a link to the documentation here*)

* [ ] Link to **testing instructions** for this release: *Enter a link to the testing instructions here, ideally in /docs/testing/releases*

* [ ] The release has a negative performance impact on sites.
    * [ ] There are new assets (JavaScript or CSS bundles)
    * [ ] There is an increase to the size of JavaScrip or CSS bundles) *please include rationale for this increase*
    * [ ] Other negative performance impacts (if yes, include list below)

* [ ] The release has positive performance impact on sites. If checked, please document these improvements here.

## Additional Notes

> This section is for additional notes related to the release.

------



###### :rocket: This pull request was generated by the [automations bot](https://github.com/woocommerce/automations) triggered by the creation of the branch: \`release/4.2.1\`. cc @Codertocat
",
    "head": "release/4.2.1",
    "owner": "nerrad",
    "repo": "tests",
    "title": "Release: 4.2.1",
  },
]
`;

exports[`branch-create-handler for release pull request generation Bails if there is no milestone for the given release branch 2`] = `
Array [
  Object {
    "body": "The release pull request has been created! This checklist is a guide to follow for the remainder of the release process. You can check off each item in this list once completed.

## Checklist

* [ ] Add changelog to readme.txt (you can get the generated changelog from the pull request description).
* [ ] Create testing notes and generate zip for testing with.
* [ ] Review checklist and add any necessary comments in the main description for the pull request.
* [ ] Build and commit GitHub tag for 4.2.1 (once pull request is approved).



###### :rocket: This comment was generated by the [automations bot](https://github.com/woocommerce/automations) triggered by the creation of the branch: \`release/4.2.1\`. cc @Codertocat
",
    "issue_number": 1235,
    "owner": "nerrad",
    "repo": "tests",
  },
]
`;

exports[`branch-create-handler for release pull request generation creates a pull request 1`] = `
Array [
  Object {
    "base": "master",
    "body": "This is the release pull request for WooCommerce Blocks plugin 3.0.0.

## Communication

>  This section is for any notes related to communicating the release Please include any extra details with each item as needed.


This release introduces:


> In this section document an overview/summary of what this release includes. You can refer to the changelog for more information


### Changelog

#### Various

- Found a bug. ([1347](https://github.com/octocat/Hello-World/issues/1347))



### Prepared Updates

The following documentation, blog posts, and changelog updates are prepared for the release:


> In this section you are highlighting all the public facing documentation that is related to the release. Feel free to remove anything that doesn't apply for this release.


**Release announcement:** *Link to release announcement post on developer.woocommerce.com (published after release)*



**Relevant developer documentation:**
_Link(s) to any developer documentation related to the release_

**Happiness Engineer:**
_Link to any special instructions or helpful notes for HE related to this release_

* [ ] The release includes a changelog entry in the readme.txt?


## Quality


> This section is for any notes related to quality around the release Please include any extra details with each item as needed. This can include notes about why something isn't checked or expanding info on your response to an item.



**For testing:**

* [ ] Changes in this release are covered by Automated Tests.

      > This section is for confirming that the release changeset is covered by automated tests. If not, please leave some details on why not and any relevant information indicating confidence without those tests.

     * [ ] Unit tests
     * [ ] E2E tests
     * [ ] for each supported WordPress and WooCommerce core versions.

* This release has been tested on the following platforms:
     * [ ] mobile
     * [ ] desktop

* [ ] This release affects public facing REST APIs.
    * [ ] It conforms to REST API versioning policy.

* [ ] This release impacts **other extensions** or **backward compatibility**.
    * [ ] The release changes the signature of public methods or functions
        * [ ] This is documented (see: *Enter a link to the documentation here*)
    * [ ] The release affects filters or action hooks.
        * [ ] This is documented (see: *Enter a link to the documentation here*)

* [ ] Link to **testing instructions** for this release: *Enter a link to the testing instructions here, ideally in /docs/testing/releases*

* [ ] The release has a negative performance impact on sites.
    * [ ] There are new assets (JavaScript or CSS bundles)
    * [ ] There is an increase to the size of JavaScrip or CSS bundles) *please include rationale for this increase*
    * [ ] Other negative performance impacts (if yes, include list below)

* [ ] The release has positive performance impact on sites. If checked, please document these improvements here.

## Additional Notes

> This section is for additional notes related to the release.

------



###### :rocket: This pull request was generated by the [automations bot](https://github.com/woocommerce/automations) triggered by the creation of the branch: \`release/3.0\`. cc @Codertocat
",
    "head": "release/3.0",
    "owner": "nerrad",
    "repo": "tests",
    "title": "Release: 3.0.0",
  },
]
`;

exports[`branch-create-handler for release pull request generation creates a pull request 2`] = `
Array [
  Object {
    "body": "The release pull request has been created! This checklist is a guide to follow for the remainder of the release process. You can check off each item in this list once completed.

## Checklist

* [ ] Add changelog to readme.txt (you can get the generated changelog from the pull request description).
* [ ] Create testing notes and generate zip for testing with.
* [ ] Review checklist and add any necessary comments in the main description for the pull request.
* [ ] Build and commit GitHub tag for 3.0.0 (once pull request is approved).



###### :rocket: This comment was generated by the [automations bot](https://github.com/woocommerce/automations) triggered by the creation of the branch: \`release/3.0\`. cc @Codertocat
",
    "issue_number": 1235,
    "owner": "nerrad",
    "repo": "tests",
  },
]
`;
